http/relay/relay_test.go

```diff
@@ -87,18 +133,100 @@ func TestRelay(t *testing.T) {
 		})
 	})
 
+	closeApp := func(t *testing.T) {
+		t.Helper()
+		r1 := numRestarts()
+		body := newRequestBody(t, "GET", "/close", nil)
+		assertString(t, "closed", body)
+
+		r2 := numRestarts()
+		assert.Equal(t, r2, r1+1)
+	}
+
+	// A bad route (such as /throw) eventually stops retrying
+	t.Run("bad route", func(t *testing.T) {
+		newH(t)
+
+		r1 := numRestarts()
+		res := httptest.NewRecorder()
+		req := httptest.NewRequest("GET", "/throw", nil)
+		h.ServeHTTP(res, req)
+		r2 := numRestarts()
+
+		// Be a good proxy now
+		assert.Equal(t, 502, res.Code)
+
+		// 3 retries == 4 tries
+		assert.Equal(t, r2-r1, 4)
+	})
+
 	t.Run("restart server", func(t *testing.T) {
+		newH(t)
+
 		t.Run("200", func(t *testing.T) {
-			res := httptest.NewRecorder()
-			req := httptest.NewRequest("GET", "/throw/env", nil)
+			body := newRequestBody(t, "GET", "/throw/env", nil)
+			assertString(t, "Hello", body)
+		})
 
-			h.ServeHTTP(res, req)
+		t.Run("use after close", func(t *testing.T) {
+			closeApp(t)
+		})
 
-			assert.Equal(t, 200, res.Code)
-			assertString(t, "Hello", res.Body.String())
+		t.Run("use after close restart loop", func(t *testing.T) {
+			for i := 0; i < 7; i++ {
+				closeApp(t)
+			}
 		})
 	})
 
+	t.Run("child process cleanup", func(t *testing.T) {
+
+		// In this test, we test that a child process who stops accepting network connections
+		// (but hasn't crashed) is gracefully asked to be shut down
```

Comment:
This is effectively the test case that I was running into in prod, `up-proxy` thought my process failed (because of a network error), however in actuality it didn't.  Therefore the process stayed around forever, and continued to use resource on the lambda container.